rclcpp/include/rclcpp/subscription_intra_process.hpp

```diff
@@ -0,0 +1,194 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__SUBSCRIPTION_INTRA_PROCESS_HPP_
+#define RCLCPP__SUBSCRIPTION_INTRA_PROCESS_HPP_
+
+#include <rmw/rmw.h>
+
+#include <functional>
+#include <memory>
+#include <utility>
+
+#include "rcl/error_handling.h"
+
+#include "rclcpp/contexts/default_context.hpp"
+#include "rclcpp/buffers/intra_process_buffer.hpp"
+#include "rclcpp/subscription.hpp"
+#include "rclcpp/type_support_decl.hpp"
+#include "rclcpp/waitable.hpp"
+
+namespace rclcpp
+{
+
+// Forward declarations
+namespace node_interfaces
+{
+class NodeTopicsInterface;
+class NodeWaitablesInterface;
+}  // namespace node_interfaces
+
+
+class SubscriptionIntraProcessBase : public rclcpp::Waitable
+{
+public:
+  RCLCPP_SMART_PTR_DEFINITIONS(SubscriptionIntraProcessBase)
+
+  SubscriptionIntraProcessBase() {}
+
+  size_t
+  get_number_of_ready_guard_conditions() {return 1;}
+
+  virtual bool
+  add_to_wait_set(rcl_wait_set_t * wait_set) = 0;
+
+  virtual bool
+  is_ready(rcl_wait_set_t * wait_set) = 0;
+
+  virtual void
+  execute() = 0;
+
+  virtual bool
+  use_take_shared_method() const = 0;
+
+  virtual const char *
+  get_topic_name() const = 0;
+
+  virtual rmw_qos_profile_t
+  get_actual_qos() const = 0;
+
+private:
+  virtual void
+  trigger_guard_condition() = 0;
+};
+
+
+template<
+  typename MessageT,
+  typename Alloc = std::allocator<void>>
+class SubscriptionIntraProcess : public SubscriptionIntraProcessBase
+{
+public:
+  RCLCPP_SMART_PTR_DEFINITIONS(SubscriptionIntraProcess)
+
+  using MessageAllocTraits = allocator::AllocRebind<MessageT, Alloc>;
+  using MessageAlloc = typename MessageAllocTraits::allocator_type;
+  using MessageDeleter = allocator::Deleter<MessageAlloc, MessageT>;
+  using ConstMessageSharedPtr = std::shared_ptr<const MessageT>;
+  using MessageUniquePtr = std::unique_ptr<MessageT, MessageDeleter>;
+
+  using BufferUniquePtr = typename intra_process_buffer::IntraProcessBuffer<MessageT>::UniquePtr;
+
+  SubscriptionIntraProcess(
+    AnySubscriptionCallback<MessageT, Alloc> callback,
+    const std::string & topic_name,
+    rmw_qos_profile_t qos_profile,
+    BufferUniquePtr buffer)
+  : any_callback_(callback), buffer_(std::move(buffer)), topic_name_(topic_name), qos_profile_(qos_profile)
+  {
+    std::shared_ptr<rclcpp::Context> context_ptr =
+      rclcpp::contexts::default_context::get_global_default_context();
+
+    rcl_guard_condition_options_t guard_condition_options =
+      rcl_guard_condition_get_default_options();
+
+    gc_ = rcl_get_zero_initialized_guard_condition();
+    rcl_ret_t ret = rcl_guard_condition_init(
+      &gc_, context_ptr->get_rcl_context().get(), guard_condition_options);
+
+    if (RCL_RET_OK != ret) {
+      throw std::runtime_error("SubscriptionIntraProcess init error initializing guard condition");
+    }
+  }
+
+  bool
+  add_to_wait_set(rcl_wait_set_t * wait_set)
```

Comment:
This method and others here, could be implemented in the base class (and live in a .cpp file).

Reply:
Created `subscription_intra_process_base.hpp`, `subscription_intra_process_base.cpp` and `subscription_intra_process.hpp`